From 633cc068e49e28bdbc2d898c20947aacc0357e71 Mon Sep 17 00:00:00 2001
From: Leo Yu-Chi Liang <ycliang@andestech.com>
Date: Tue, 13 Aug 2024 15:46:23 +0800
Subject: [PATCH 112/168] smu: andes: irq/irq-sifive-plic: Add plic_set_wake to
 support wake event (#164)

We need irq_set_wake hook to get wake event information,
so revert commit f5259b045c19 ("smu: andes: Revert "irqchip/sifive-plic: Support wake IRQs")

When entering suspend, kernel PM framework will check all the drivers
to see if they have the ability to wake up the system.

Implement plic_set_wake to acquire wake event information.

Signed-off-by: Leo Yu-Chi Liang <ycliang@andestech.com>
---
 drivers/irqchip/irq-sifive-plic.c | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/drivers/irqchip/irq-sifive-plic.c b/drivers/irqchip/irq-sifive-plic.c
index bf0b40b0fad4..9ccf989b945c 100644
--- a/drivers/irqchip/irq-sifive-plic.c
+++ b/drivers/irqchip/irq-sifive-plic.c
@@ -63,6 +63,8 @@
 
 #define PLIC_QUIRK_EDGE_INTERRUPT	0
 
+unsigned long *andes_wake_event;
+
 struct plic_priv {
 	struct cpumask lmask;
 	struct irq_domain *irqdomain;
@@ -157,6 +159,16 @@ static void plic_irq_eoi(struct irq_data *d)
 	}
 }
 
+static int plic_set_wake(struct irq_data *d, unsigned int on)
+{
+	if (on)
+		__assign_bit(d->hwirq, andes_wake_event, true);
+	else
+		__assign_bit(d->hwirq, andes_wake_event, false);
+
+	return 0;
+}
+
 #ifdef CONFIG_SMP
 static int plic_set_affinity(struct irq_data *d,
 			     const struct cpumask *mask_val, bool force)
@@ -193,12 +205,12 @@ static struct irq_chip plic_edge_chip = {
 	.irq_ack	= plic_irq_eoi,
 	.irq_mask	= plic_irq_mask,
 	.irq_unmask	= plic_irq_unmask,
+	.irq_set_wake	= plic_set_wake,
 #ifdef CONFIG_SMP
 	.irq_set_affinity = plic_set_affinity,
 #endif
 	.irq_set_type	= plic_irq_set_type,
-	.flags		= IRQCHIP_SKIP_SET_WAKE |
-			  IRQCHIP_AFFINITY_PRE_STARTUP,
+	.flags		= IRQCHIP_AFFINITY_PRE_STARTUP,
 };
 
 static struct irq_chip plic_chip = {
@@ -208,12 +220,12 @@ static struct irq_chip plic_chip = {
 	.irq_mask	= plic_irq_mask,
 	.irq_unmask	= plic_irq_unmask,
 	.irq_eoi	= plic_irq_eoi,
+	.irq_set_wake	= plic_set_wake,
 #ifdef CONFIG_SMP
 	.irq_set_affinity = plic_set_affinity,
 #endif
 	.irq_set_type	= plic_irq_set_type,
-	.flags		= IRQCHIP_SKIP_SET_WAKE |
-			  IRQCHIP_AFFINITY_PRE_STARTUP,
+	.flags		= IRQCHIP_AFFINITY_PRE_STARTUP,
 };
 
 static int plic_irq_set_type(struct irq_data *d, unsigned int type)
@@ -422,6 +434,10 @@ static int __init __plic_init(struct device_node *node,
 	struct plic_handler *handler;
 	unsigned int cpu;
 
+	andes_wake_event = kzalloc(sizeof(unsigned long), GFP_KERNEL);
+	if (WARN_ON(!andes_wake_event))
+		return -ENOMEM;
+
 	priv = kzalloc(sizeof(*priv), GFP_KERNEL);
 	if (!priv)
 		return -ENOMEM;
-- 
2.34.1

